boards/posix/native_posix/native_posix.dts

```diff
@@ -6,6 +6,7 @@
 
 /dts-v1/;
 #include <posix/posix.dtsi>
+#include <dt-bindings/gpio/gpio.h>
```

Comment:
> OK, this got long, so I'm stopping now. Not everything here is "change requested", but a lot is.

No problem - I'll make the requested changes when I have some time. I hope you don't mind that I keep the PR open for a while.

> At a high level I'm concerned about simulator fidelity and validation. That in-tree GPIO tests pass with this is insufficient: those tests were written based on an assumption that hardware GPIO peripherals behave correctly if the configuration is correct.

Yeah, that was mainly intended to keep things simple at first. E.g. the assumption was made that inputs have pull-down resistors, but that's obviously not always the case. Similarly, went with the idea of GPIO registers holding values for input or output, but not both, but that's not the case either in a lot of hardware. Do you think it's better to use a `gpio_port_pins_t` for input, output & hi-z? I'm just wondering because maybe I can store all pin attributes in a 32-bit value (like I do for most of the GPIO configuration). 

Speaking of GPIO configuration, I think it would be pretty useful to add an API call for querying that on a per-pin basis - kind of the inverse of `gpio_config()`. It would be extremely useful in my application, and as you can see, I needed to expose that functionality just to get GPIO simulation to work with "wiring".

> The simulator should have its own dedicated test, preferably one run with coverage enabled, that checks things like whether changing GPIO configuration away from and back to a particular mode inadvertently clears state that would be persisted in real hardware.

Sure - how do I go about enabling coverage? Can you point me to an example? If you're more specific about the state that should be persisted, I can add tests for that.

> Examples that I don't think this does quite right include the input/output/hi-Z state of a pin; handling pull configurations; and whether an interrupt event captured by the hardware but ignored due to interrupts being enabled is triggered immediately if the pin configuration is changed.

Agreed. I'll look at in/out/hi-z. 

I might need you to go into specifics for the latter part of that though. No rush, just whenever you have some time.

> I would really like to see a documented shared understanding of the role simulation will play in Zephyr evolution, so we don't have conflicting models. This one seems to have no external dependencies, and could be run on real microcontroller hardware (e.g. to simulate an external device that isn't available in hardware). The one originally proposed in #26003 before it got derailed might be able to do more by being "limited" to a single platform that's pretty powerful. Or maybe they're entirely compatible and will work together.
> 
> If we don't coordinate things, that's less likely to be true.

Sure. My main purpose is to be able to "mock out" a GPIO controller so that I can create integration and system tests that can run in CI. In the end, we're just dealing with information, and I'm particularly focused on the communication of that information at the moment, and not whether the "mock" (model) is perfect. So the device-under-test in my case is a communications API that is designed specifically for dealing with embedded I/O. However, I think it's important to make that model perfect (or as close to perfect as can be expected) in the long run. Certainly, at some point, there will be a different framework that will interact with I/O and it would most likely benefit by being able to mock out I/O.

If there is a different model that this could be useful for, I'd be happy to hear about that as well.